pkg/liquidity-mining/test/BalancerMinter.test.ts

```diff
@@ -1,25 +1,249 @@
+        });
+      });
+
+      context('when caller is allowed', () => {
+        sharedBeforeEach(async () => {
+          await minterContract.connect(user).setMinterApproval(minter.address, true);
+        });
+
+        itCallsVirtualHooks();
+      });
+    });
+
+    describe('mint_for', () => {
+      sharedBeforeEach(async () => {
+        mintFunction = async () => await (await minterContract.connect(minter).mint_for(gauge, user.address)).wait();
+        mintStaticFunction = async () => await minterContract.connect(minter).callStatic.mint_for(gauge, user.address);
+      });
+
+      context('when the caller is not allowed', () => {
+        it('does not call _mintFor virtual hook', async () => {
+          const receipt = await mintFunction();
+          expectEvent.notEmitted(receipt, 'MintFor');
+        });
+      });
+
+      context('when caller is allowed', () => {
+        sharedBeforeEach(async () => {
+          await minterContract.connect(user).setMinterApproval(minter.address, true);
+        });
+
+        it('calls internal _mintFor virtual hook', async () => {
+          const receipt = await mintFunction();
+          expectEvent.inReceipt(receipt, 'MintFor', { gauge, user: user.address });
+        });
+
+        it('returns nothing', async () => {
+          expect(await mintStaticFunction()).to.be.deep.eq([]);
+        });
+      });
+    });
+  });
+
+  describe('mintMany / mintManyFor', () => {
+    let mintManyFunction: () => Promise<ContractReceipt>;
+    let mintManyStaticFunction: () => Promise<BigNumber>;
+    let gauges: string[];
+
+    const expectedMintManyForReturn = 83;
+
+    function itCallsVirtualHooks() {
+      it('calls internal _mintForMany virtual hook', async () => {
+        const receipt = await mintManyFunction();
+        expectEvent.inReceipt(receipt, 'MintForMany', { gauges, user: user.address });
+      });
+
+      it('returns the same value as the internal _mintForMany virtual hook', async () => {
+        expect(await mintManyStaticFunction()).to.be.eq(expectedMintManyForReturn);
+      });
+    }
+
+    sharedBeforeEach(async () => {
+      gauges = await Promise.all(range(5).map(randomAddress));
+      await minterContract.setMockMintForMany(expectedMintManyForReturn);
+    });
+
+    describe('mintMany', () => {
+      sharedBeforeEach(async () => {
+        mintManyFunction = async () => await (await minterContract.connect(user).mintMany(gauges)).wait();
+        mintManyStaticFunction = async () => await minterContract.connect(user).callStatic.mintMany(gauges);
+      });
+
+      itCallsVirtualHooks();
+    });
+
+    describe('mintManyFor', () => {
+      sharedBeforeEach(async () => {
+        mintManyFunction = async () =>
+          await (await minterContract.connect(minter).mintManyFor(gauges, user.address)).wait();
+        mintManyStaticFunction = async () =>
+          await minterContract.connect(minter).callStatic.mintManyFor(gauges, user.address);
+      });
```

Comment:
You could also test `user` calling these if you wanted

Reply:
actually... it's the same case.
If user calls `mintFor` for themselves, it'll revert because by default it's not allowed to. In other words, to call `mintFor` you'd need to first call `setMinterApproval` for yourself.

We could correct that by changing the require to
```
        require(_allowedMinter[msg.sender][user] || msg.sender == user, "Caller not allowed to mint for user");
```